### PR TITLE
Only Show Runs in the Data Cache When It is Available

### LINKED ISSUE #36860: ISIS Reflectometry GUI auto-processing stops if file cannot be found
**Original reporter:** Stephen H at ISIS on behalf of Reflectometry Group

**Describe the bug**

The ISIS Reflectometry GUI auto-process feature looks up new runs in the journal viewer and then automatically transfers them to the main table on the runs tab and reduces them. This works fine as long as the run file that was returned in the search is available for the reduction that will follow. If it's not then a file not found error is logged and the row reduction fails. The GUI does not appear to attempt to try and reduce the row again.

This never previously caused problems because new runs only appeared in the journal viewer when the run file was available in the archive. With recent changes to IDAaaS, runs now seem to appear in the journal viewer before they are available in the instrument data cache. This means that users who don't or can't mount the archive encounter this issue when trying to auto-process new data.

**To Reproduce**

To reproduce properly would require you to be auto-processing a currently running experiment on IDAaaS without the archive mounted. You would need permissions to access the experimental data and have the folder in your Mantid user directories.

To simulate you can do the following:

1) Find some test runs for a given experiment.
1) Put some but not all into a folder on your local machine.
1) Turn off the archive in Mantid and open the ISIS Reflectometry interface.
1) On the Runs tab, enter the experiment RB number into the `Investigation ID` field and enter the cycle into the `Cycle` field. Click the Autoprocess button.
1) The search result should populate with the full list of runs for the experiment. Only the files that can be found will reduce and the rest will result in errors. Adding the missing data files into the folder afterwards will not result in them being found and reduced.

**Expected behavior**

The auto-processing should be able to handle a delay in the files becoming available. There are a few ways this could be done, but as a start-point we could investigate preventing runs found in the search from being added to the search results until the file is available. We could log at debug level if a run has been found when the corresponding file cannot be accessed. Once the file is available, adding it to the search results at that point should in theory automatically trigger the rest of the process.

**Platform/Version (please complete the following information):**
 - Has been an issue most likely from the introduction of the feature.

**Additional context**

From investigations so far, the auto-process feature works as follows:

- Clicking the Autoprocess button results in a call to `QtRunsView::on_actionAutoreduce_triggered`. This goes and checks for new runs, which stops any polling for new runs while it's doing this.

- After getting search results, any new runs are added to the table in `QtCatalogSearcher::convertJournalResultsTableToSearchResults`.

- This all happens asynchronously and when the search and the reduction have finished there is a call to `RunsPresenter::autoreductionCompleted()`. This starts the polling again, which makes a new call to `RunsPresenter::notifyCheckForNewRuns()` every 5 seconds to check for new runs.

### PR DESCRIPTION
**Pull request by cailafinn**
#### Summary of work
When the Data Cache is available (such as on IDAaaS), only the runs that are available in it will be shown in the investigation menu, and therefore be available for transfer and auto-processing. 

This is to account for two problems:
1. Not all of the files in the archive are available in the data cache (yet).
2. Items are added to the archive (which is what we query with the invesigation ID to get the runs list) _before_ the runs are available in the data cache. 

Making sure only the items in the data cache before showing them to auto-processing ensures that we don't just instantly fail by not finding the file. 

Fixes #36860 

#### Further detail of work

- Adds some functionality to the ISISInstrumentDataCache class to check if the cache is available and to get the list of runs available within it. 
- Makes a change to the conversion to the search results generator, so that it filters out any of the runs that aren't present in the data cache. 

### To test:
The easiest way to do this is probably to just make a package build to use on IDAaaS, but if you want to test it locally, here you go (you'll need access to the archive or at least a decent number of INTER files with the same investigation ID):

1. Boot workbench and open the ISIS Reflectometry Interface
2. Enter 1120015 and 11_3
3. Use Ctrl+Shift to select a bunch of files and transfer them to the runs table
4. Process the files and save out the workspaces from the TOF group.
5. In your file browser rename the files to INTER000[11991].nxs (replacing the run number with the correct one)
6. Pick a location on your machine for your fake data cache. 
7. Have the files in a directory named after the instrument. `INTER` in this case. 
8. In the `INTER` directory, create an `INTER_index.json` file containing the following (making the run numbers match, obviously): 
```json
{
   "11990": "11_3",
    "11991": "11_3",
    "11992": "11_3",
    "11993": "11_3",
    "11994": "11_3",
    "HIDE_11995": "11_3",
    "11996": "11_3",
    "11997": "11_3",
    "11998": "11_3",
    "HIDE_11999": "11_3",
    "12000": "11_3",
    "12001": "11_3",
    "12002": "11_3",
    "12003": "11_3",
    "12004": "11_3"
}
```
10. in the end, your structure should look like this:
image

11. Final bit of setup, set the datacache location in your `mantid.user.properties` file: `datacachesearch.directory=/path/to/FakeCache`
12. Then restart mantid to pick up the change to the properties
13. Open the Refl interface
14. Enter your run number and cycle from before
15. Click autoprocess
16. Only the files present in the index file should be moved over to the runs table
17. Edit the index file to remove the `HIDE` parts.
18. The autoprocessing should pick up the newly added runs and process them automatically. 

---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

### CODE REVIEW

**Comment by jclarkeSTFC** [comment]
When it's finished the package will be here (just making a note for myself): https://builds.mantidproject.org/job/build_packages_from_branch/810/

File: Framework/API/src/ISISInstrumentDataCache.cpp [hunk 1]
```diff
@@ -45,8 +41,40 @@ std::string Mantid::API::ISISInstrumentDataCache::getFileParentDirectoryPath(con
...
+ */
+std::pair<std::string, Json::Value>
+ISISInstrumentDataCache::openCacheJsonFile(std::string const &instrumentName) const {
+  // Open index json file
+  std::filesystem::path const &jsonPath =
+      std::filesystem::path(m_dataCachePath) / instrumentName / (instrumentName + "_index.json");
+  std::ifstream ifstrm{jsonPath};
+  if (!ifstrm) {
+    throw std::invalid_argument("Could not open index file: " + jsonPath.string());
+  }
```
Comment (jclarkeSTFC) [inline Framework/API/src/ISISInstrumentDataCache.cpp:68]: It looks like the old code attempted to use the short name of the instrument if the full name didn't work, does it not need to do that any more?

Comment (cailafinn) [reply Framework/API/src/ISISInstrumentDataCache.cpp:68]: Ah, I think (very much not certain) that this was added later and so got missed during a rebase. Good catch, thanks

**Review by jclarkeSTFC — DISMISSED** [review]
Test worked fine, code looks good. I tested with a fake data cache on Ubuntu because I don't have access to the INTER data cache on IDAaaS.

**Review by rbauststfc — CHANGES_REQUESTED** [review]
We've discussed that we need to consider the case where an instrument scientist has the archive mounted. If archive access is available then we shouldn't filter the search results.

**Review by jclarkeSTFC — DISMISSED** [review]
With the archive off, only the runs in the data cache are shown (I used a local fake cache), un-hiding a run using the `json` worked fine, autoprocessing picked it up. With the archive on I saw all of the runs (including ones that aren't in the cache) in the list on the left of the reflectometry interface.

**Review by rbauststfc — CHANGES_REQUESTED** [review]
The solution to ensure results are returned from the archive when it's switched on looks really great, and also really good to add something to the documentation about it. To support the testing done already, I've checked on IDAaaS that older runs that are no longer in the data cache are not picked up unless the archive is switched on in Mantid, and that's all looking good. I also have a couple of RB folders that I have permissions for on IDAaaS, so I was able to test auto-processing with those and all looked good there too.

There were just a couple of things that I noticed about documentation that might be good to address before we merge this in.

**Review by rbauststfc — APPROVED** [review]
Thanks for the documentation changes, that's all looking really good now. That was the only change since James' review, and everything else was looking good then, so I think this is ready to be gatekeepered.

**Comment by rbauststfc** [comment]
I've removed James as an assignee so it's clear we're looking for a gatekeeper for this one.

**Pull request merged**